Detectors/AOD/src/AODProducerWorkflowSpec.cxx

```diff
@@ -1541,6 +1580,299 @@ void AODProducerWorkflowDPL::run(ProcessingContext& pc)
+{
+  // Orphan tracks need to refer to some globalBC and for tracks with TOF this BC should be whithin an orbit
+  // from the track abs time (to guarantee time precision). Therefore, we may need to insert some dummy globalBCs
+  // to guarantee proper reference.
+  // complete globalBCs by dummy entries necessary to provide BC references for TOF tracks with requested precision
+  // to provide a reference for the time of the orphan tracks we should make sure that there are no gaps longer
+  // than needed to store the time with sufficient precision
+
+  // estimate max distance in BCs between TOF time and eventual reference BC
+  int nbitsFrac = 24 - (32 - o2::math_utils::popcount(mTrackTime)); // number of bits used to encode the fractional part of float truncated by the mask
+  int nbitsLoss = std::max(0, int(std::log2(TOFTimePrecPS)));       // allowed bit loss guaranteeing needed precision in PS
+  assert(nbitsFrac > 1);
+  std::uint64_t maxRangePS = std::uint64_t(0x1) << (nbitsFrac + nbitsLoss);
+  int maxGapBC = maxRangePS / (o2::constants::lhc::LHCBunchSpacingNS * 1e3); // max gap in BCs allowing to store time with required precision
+  LOG(info) << "Max gap of " << maxGapBC << " BCs to closest globalBC reference is needed for TOF tracks to provide precision of "
+            << TOFTimePrecPS << " ps";
+
+  // check if there are tracks orphan tracks at all
+  if (!trackRef.getEntries()) {
+    return;
+  }
+  // the bscMap has at least TF start BC
+  std::uint64_t maxBC = mStartIR.toLong();
+  const auto& tofClus = data.getTOFClusters();
+  for (int src = GIndex::NSources; src--;) {
+    if (!GIndex::getSourceDetectorsMask(src)[o2::detectors::DetID::TOF]) { // check only tracks with TOF contribution
+      continue;
+    }
+    int start = trackRef.getFirstEntryOfSource(src);
+    int end = start + trackRef.getEntriesOfSource(src);
+    for (int ti = start; ti < end; ti++) {
+      auto& trackIndex = GIndices[ti];
+      const auto& tofMatch = data.getTOFMatch(trackIndex);
+      const auto& tofInt = tofMatch.getLTIntegralOut();
+      float intLen = tofInt.getL();
+      float tofExpMom = 0.;
+      if (tofInt.getTOF(o2::track::PID::Pion) > 0.f) {
+        float expBeta = (intLen / (tofInt.getTOF(o2::track::PID::Pion) * cSpeed));
+        tofExpMom = o2::constants::physics::MassPionCharged * expBeta / std::sqrt(1.f - expBeta * expBeta);
+      } else {
+        continue;
+      }
+      double massZ = o2::track::PID::getMass2Z(data.getTrackParam(trackIndex).getPID());
+      double energy = sqrt((massZ * massZ) + (tofExpMom * tofExpMom));
+      double exp = intLen * energy / (cSpeed * tofExpMom);
+      auto tofSignal = (tofMatch.getSignal() - exp) * 1e-3; // time in ns wrt TF start
+      auto bc = relativeTime_to_GlobalBC(tofSignal);
+
+      auto it = bcsMap.lower_bound(bc);
+      if (it == bcsMap.end() || it->first > bc + maxGapBC) {
+        bcsMap.emplace_hint(it, bc, 1);
```

Comment:
So you now add the dummy one closest to the tof time, right?
An alternative would be to use a bc number which actually had a valid bunch crossing. So all BCs would sit where we were sure true bunch crossings have taken place (even if it did not trigger any data taking).

Reply:
This will not work for the cosmics. Also, for the beam with a few BCs only it is not guaranteed that you can find a BC with the `maxGapBC`. In any case, this fake BC is merely a reference.

Reply:
I thought 1 BC per turn gives us already 4.5 ps precision and here you aim at 16ps?
Yes, cosmics, needs to be treated differently.

Reply:
Wth the maxGap=orbit and for the non-truncated float the precision in ps is indeed 
```
88924.e3/(1<<24)
(double) 5.3002834
```
but with default truncation of 8 lowest bits it is not enough:
```
88924.e3/(1<<16)
(double) 1356.8726
```
So, I am calculating the maxGap depending on the requested precision and adding fake BC to respect it.